File: FestivalJava/Networking/src/main/java/rpcprotocol/FestivalClientRpcWorker.java
package rpcprotocol;

import AppServices.IFestivalClient;
import AppServices.IFestivalServer;
import Domain.Artist;
import Domain.Show;
import Domain.User;
import Validation.Exceptions.ServiceException;
import Validation.Exceptions.ValidatorException;
import dto.*;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.Socket;
import java.util.List;

/**
 * Created by Sebi on 29-Mar-17.
 */
public class FestivalClientRpcWorker implements Runnable, IFestivalClient {
    //server
    private IFestivalServer server;

    //connection
    private Socket connection;

    //input + output streams
    private ObjectInputStream inputStream;
    private ObjectOutputStream outputStream;

    //indicates connection state
    private volatile boolean connected;

    //constructor
    public FestivalClientRpcWorker(IFestivalServer server, Socket connection){
        this.server = server;
        this.connection = connection;

        try{
            outputStream = new ObjectOutputStream(connection.getOutputStream());
            outputStream.flush();
            inputStream = new ObjectInputStream(connection.getInputStream());
            connected = true;
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void run() {
        while(connected){
            try{
                //get a request
                Object request = inputStream.readObject();
                System.out.println("Got a request");

                //process it
                Response response = handleRequest((Request) request);
                System.out.println("Processed request");

                //send response if able
                if (response != null){
                    sendResponse(response);
                    System.out.println("Sent response");
                }


            } catch (IOException | ClassNotFoundException ex){
                System.out.println("Disconnected while trying to read");
                break;
            }
        }

        try{
            inputStream.close();
            outputStream.close();
            connection.close();
        } catch (IOException ex){
            ex.printStackTrace();
        }
    }

    //handles a request
    private Response handleRequest(Request request){
        Response response = null;

        if (request.getType() == RequestType.LOGIN){
            System.out.println("Login request");

            UserDTO userDTO = (UserDTO)request.getData();
            User user = DTOUtils.getUserFromDTO(userDTO);

            try{
                server.login(user, this);
                System.out.println("Login success");
                return new Response.Builder().type(ResponseType.OK).build();
            } catch (ServiceException e) {
                System.out.println("Login error");
                return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
            }
        }

        if (request.getType() == RequestType.LOGOUT){
            System.out.println("Logout request");

            UserDTO userDTO = (UserDTO)request.getData();
            User user = DTOUtils.getUserFromDTO(userDTO);

            try{
                server.logout(user, this);
                connected = false;
                return new Response.Builder().type(ResponseType.OK).build();
            } catch (ServiceException e){
                return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
            }
        }

        if (request.getType() == RequestType.GET_ARTISTS){
            System.out.println("Get artists request");

            try {
                List<Artist> artists = server.getArtists();
                List<ArtistDTO> artistDTOs = DTOUtils.getListArtistDTO(artists);
                return new Response.Builder().type(ResponseType.GET_ARTISTS).data(artistDTOs).build();
            } catch (ServiceException e) {
                return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
            }
        }

        if (request.getType() == RequestType.GET_SHOWS_FOR_ARTIST){
            System.out.println("Get shows for artists request");

            try{
                Integer idArtist = (Integer)request.getData();
                List<Show> shows = server.getShowsForArtist(idArtist);
                List<ShowDTO> showDTOs = DTOUtils.getListShowDTO(shows);
                return new Response.Builder().type(ResponseType.GET_SHOWS).data(showDTOs).build();
            } catch (NumberFormatException e){
                return new Response.Builder().type(ResponseType.ERROR).data("Id must be integer").build();
            } catch (ServiceException e) {
                return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
            }
        }

        if (request.getType() == RequestType.GET_SHOWS_FOR_DATE){
            System.out.println("Get shows for date request");

            try{
                String date = (String) request.getData();
                List<Show> shows = server.getShowsForDate(date);
                List<ShowDTO> showDTOs = DTOUtils.getListShowDTO(shows);
                return new Response.Builder().type(ResponseType.GET_SHOWS).data(showDTOs).build();
            } catch (ServiceException e) {
                return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
            }
        }

        if (request.getType() == RequestType.BUY_TICKETS){
            System.out.println("Buy tickets request");

            BuyTicketsDTO buyTicketsDTO = (BuyTicketsDTO) request.getData();
            try {
                server.buyTicketsForShow(buyTicketsDTO.getIdShow(), buyTicketsDTO.getClientName(),
                        buyTicketsDTO.getNumberOfTickets(), buyTicketsDTO.getUsername());
                return new Response.Builder().type(ResponseType.OK).build();
            } catch (ServiceException | ValidatorException e) {
                return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
            }
        }

        return response;
    }

    //sends a response
    private void sendResponse(Response response) throws IOException{
        System.out.println("Sending response " + response);
        outputStream.writeObject(response);
        System.out.println("flushing");
        outputStream.flush();
    }

    @Override
    public void showUpdated(Show show) throws ServiceException {
        ShowDTO showDTO = DTOUtils.getShowDTO(show);
        Response response = new Response.Builder().type(ResponseType.SHOWS_UPDATED).data(showDTO).build();
        try {
            sendResponse(response);
        } catch (IOException e) {
            throw new ServiceException("Update notify send response error");
        }
    }
}

File: FestivalC#/Festival/Networking/rpcprotocol/FestivalClientRpcWorker.cs
using Festival.Service;
using System;
using Festival.Model;
using Festival.Service.AppServices;
using System.Net.Sockets;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using Networking.dto;
using Services.Validation.Exceptions;
using System.Collections.Generic;

namespace Networking.rpcprotocol
{
    public class FestivalClientRpcWorker : IFestivalClient
    {
        //server
        private IFestivalServer server;

        //connection to client
        private TcpClient connection;

        //object formatter
        private IFormatter formatter;

        //object stream
        private NetworkStream stream;

        //connection status
        private volatile bool connected;

        public FestivalClientRpcWorker(IFestivalServer server, TcpClient connection)
        {
            this.server = server;
            this.connection = connection;
            try
            {
                stream = connection.GetStream();
                formatter = new BinaryFormatter();
                connected = true;
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        public virtual void run()
        {
            while (connected)
            {
                try
                {
                    //read request
                    object request = formatter.Deserialize(stream);

                    //handle request
                    Response response = handleRequest((Request)request); 

                    //send response
                    if (response != null)
                    {
                        sendResponse(response);
                    }
                }
                catch (Exception ex)
                {
                    Console.WriteLine(ex.Message);
                    break;
                }
            }

            try
            {
                stream.Close();
                connection.Close();
            }
            catch (Exception e)
            {
                Console.WriteLine(e.Message);
            }
        }


        public void showUpdated(Show show)
        {
            ShowDTO showDTO = DTOUtils.getShowDTO(show);
            Response response = new Response(ResponseType.SHOWS_UPDATED, showDTO);
            try
            {
                sendResponse(response);
            }
            catch (Exception e)
            {
                throw new ServiceException("Update notify send response error");
            }
        }


        private Response handleRequest(Request request)
        {
            if (request.Type == RequestType.LOGIN)
            {
                Console.WriteLine("Login request ...");
                UserDTO userDTO = (UserDTO)request.Data;
                User user = DTOUtils.getUserFromDTO(userDTO);
                try
                {
                    lock (server)
                    {
                        server.login(user, this);
                    }
                    return new Response(ResponseType.OK, "ok");
                }
                catch (ServiceException ex)
                {
                    return new Response(ResponseType.ERROR, ex.Message);
                }
            }

            if (request.Type == RequestType.LOGOUT)
            {
                Console.WriteLine("Logout request");
                UserDTO userDTO = (UserDTO)request.Data;
                User user = DTOUtils.getUserFromDTO(userDTO);
                try
                {
                    lock (server)
                    {
                        server.logout(user, this);
                    }
                    connected = false;
                    return new Response(ResponseType.OK, "ok");
                } 
                catch(ServiceException ex)
                {
                    return new Response(ResponseType.ERROR, ex.Message);
                }
            }

            if (request.Type == RequestType.GET_ARTISTS)
            {
                Console.WriteLine("Artist list request");
               
                try
                {
                    List<Artist> artists = new List<Artist>();

                    lock (server)
                    {
                        artists = server.getArtists();
                    }
                    return new Response(ResponseType.OK, DTOUtils.getListArtistDTO(artists));
                }
                catch (ServiceException ex)
                {
                    return new Response(ResponseType.ERROR, ex.Message);
                }
            }

            if (request.Type == RequestType.GET_SHOWS_FOR_ARTIST)
            {
                Console.WriteLine("Shows for artist request");

                try
                {
                    int idArtist = (Int32)request.Data;

                    List<Show> shows = new List<Show>();

                    lock (server)
                    {
                        shows = server.getShowsForArtist(idArtist);
                    }
                    return new Response(ResponseType.OK, DTOUtils.getListShowDTO(shows));
                }
                catch (ServiceException ex)
                {
                    return new Response(ResponseType.ERROR, ex.Message);
                }
            }

            if (request.Type == RequestType.GET_SHOWS_FOR_DATE)
            {
                Console.WriteLine("Shows for date request");

                try
                {
                    string date = (string)request.Data;

                    List<Show> shows = new List<Show>();

                    lock (server)
                    {
                        shows = server.getShowsForDate(date);
                    }
                    return new Response(ResponseType.OK, DTOUtils.getListShowDTO(shows));
                }
                catch (ServiceException ex)
                {
                    return new Response(ResponseType.ERROR, ex.Message);
                }
            }

            if (request.Type == RequestType.BUY_TICKETS)
            {
                Console.WriteLine("Buy tickets request");

                try
                {
                    BuyTicketsDTO buyTicketsDTO = (BuyTicketsDTO)request.Data;

                    lock (server)
                    {
                        server.buyTicketsForShow(buyTicketsDTO.IdShow,
                            buyTicketsDTO.ClientName,
                            buyTicketsDTO.NumberOfTickets,
                            buyTicketsDTO.Username);
                    }
                    return new Response(ResponseType.OK, "ok");
                }
                catch (ServiceException ex)
                {
                    return new Response(ResponseType.ERROR, ex.Message);
                }
            }

            return null;
        }

        //send response over network
        private void sendResponse(Response response)
        {
            Console.WriteLine("sending response " + response);
            formatter.Serialize(stream, response);
            stream.Flush();
        }
    }
}

